### PR TITLE
css support for style attribute name

### PR DESCRIPTION
**Pull request by neraliu**
@adon, would u please help to review this code change for css support in the style attribute name. I am planning to introduce 4 new types of css filters as follow, please comment.

ycufull - the filter for url() pattern for CSS.
yced - the filter for double quoted string in CSS.
yces - the filter for single quoted string in CSS.
yceu - the filter for ident in CSS. (note: ident is defined in the spec)

reference:
http://www.w3.org/TR/CSS21/grammar.html

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: src/context-parser-handlebars.js [hunk 1]
```diff
@@ -54,6 +58,22 @@ var reURIContextStartWhitespaces = /^(?:[\u0000-\u0020]|&#[xX]0*(?:1?[1-9a-fA-F]
...
+/*
+'&#0*32;?|&#[xX]0*20;?|&#0*9;?|&#[xX]0*9;?|&Tab;|&#0*10;?|&#[xX]0*[aA];?|&NewLine;|&#0*12;?|&#[xX]0*[cC];?|&#0*13;?|&#[xX]0*[dD];?|\t|\r|\n|\f'; // space,\t,\r,\n,\f
+'&#0*58;?|&#[xX]0*3[aA];?|&colon;'      // colon
+'&#0*59;?|&#[xX]0*3[bB];?|&semi;'       // semicolon
+'&#0*40;?|&#[xX]0*28;?|&lpar;'          // (
+'&#0*41;?|&#[xX]0*29;?|&rpar;'          // )
+*/
+var cssReplaceChar = [ ' ', ':', ';', '(', ')' ];
+var reCss = [
+    /&#0*32;?|&#[xX]0*20;?|&#0*9;?|&#[xX]0*9;?|&Tab;|&#0*10;?|&#[xX]0*[aA];?|&NewLine;|&#0*12;?|&#[xX]0*[cC];?|&#0*13;?|&#[xX]0*[dD];?|\t|\r|\n|\f/g,
```
Comment (adon-at-work) [inline src/context-parser-handlebars.js]: how about \v (i.e. &#11; or &#0B;) in IE? have you tested its behavior?

Comment (adon-at-work) [inline src/context-parser-handlebars.js]: and it seems to me you missed matching the raw ` ` character

File: src/context-parser-handlebars.js [hunk 2]
```diff
@@ -604,11 +624,69 @@ ContextParserHandlebars.prototype.addFilters = function(stateObj, input) {
...
+                        attributeValue = attributeValue.replace(reCss[j], cssReplaceChar[j]);
+                    }
+                    attributeValue = attributeValue.replace(/ /g, ''); // ok to remove all space and space has been html decoded.
+                    debug("addFilter:attributeValue:["+attributeValue+"]");
+
+                    /* split the string as per http://www.w3.org/TR/css-style-attr/ with ':' and ';' */
+                    var kv = attributeValue.split(';'); // it will return new array even there is no ';' in the string
+                    var v = kv[kv.length-1].split(':'); // only handling the last element
+
+                    if (v.length && v.length === 2) {
```
Comment (adon-at-work) [inline src/context-parser-handlebars.js]: checking `v.length` here is redundunt

File: src/context-parser-handlebars.js [hunk 3]
```diff
@@ -604,11 +624,69 @@ ContextParserHandlebars.prototype.addFilters = function(stateObj, input) {
...
+                        filters = [];
+
+                        var prop = v[0],
+                            expr = v[1];
+                        debug("addFilter:prop:"+prop+",expr:"+expr);
+
+                        /* TODO: we can whitelist the property here */
+                        if (prop !== '') { // && whitelisted
```
Comment (adon-at-work) [inline src/context-parser-handlebars.js]: I perceive something like below:

```javascript
// the code below involves less decoding than the PR
var lastPair = attributeValue.split(/;|&#0*58;?|&#[xX]0*3[aA];?|&colon;/g).pop().split(/:|&#0*59;?|&#[xX]0*3[bB];?|&semi;/g);

if (lastPair.length === 2) {
    var prop = lastPair[0], expressions;
    if (prop) { // && whitelisted
        expressions = lastPair[1].split(reCSSWhiteSpaces);
        for (var i = 0, len = expressions.length, expr; i < len; i++) {
            expr = expressions[i];
            if (expr === ‘’) continue;
            //...
        }
    }
}
```

File: src/context-parser-handlebars.js [hunk 4]
```diff
@@ -604,11 +624,69 @@ ContextParserHandlebars.prototype.addFilters = function(stateObj, input) {
...
+
+                            /* space has been trimmed above */
+                            if (expr.match(/^url\($/)) {
```
Comment (adon-at-work) [inline src/context-parser-handlebars.js]: you need a regexp that also match url case insensitively, and the html encoded versions of `u` `r` `l`

Comment (adon-at-work) [inline src/context-parser-handlebars.js]: something like this, please double check
`var reCSSUrlBracket = /^(?:[uU]|&#0*(?:117|85);?|&#[xX]0*[57]5;?)(?:[rR]|&#0*(?:114|82);?|&#[xX]0*[57]2;?)(?:[lL]|&#0*(?:107|75);?|&#[xX]0*[46][Bb];?)(?:\?|&#0*40;?|&#[xX]0*28;?|&lpar;)$/i;

File: tests/samples/files/handlebarsjs_filter_attr_value_style_003.hbs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+<div style='color: {{#if xxx}} {{color3}} {{else}} {{color4}} {{/if}};'>
+<div style="color: {{#if xxx}} '{{color5}}' {{else}} '{{color6}}' {{/if}};">
+<div style='color: {{#if xxx}} "{{color7}}" {{else}} "{{color8}}" {{/if}};'>
+<div style=color:{{#if xxx}}{{color9}}{{else}}{{color10}}{{/if}};>
+
+<div style='color: {{#if xxx}} '{{invalid1}}' {{else}} '{{invalid2}}' {{/if}};'>
+<div style="color: {{#if xxx}} "{{invalid3}}" {{else}} "{{invalid4}}" {{/if}};">
+
+<div style=color:{{#if xxx}}"{{invalid5}}"{{else}}"{{invalid6}}"{{/if}};">
+<div style=color:{{#if xxx}}'{{invalid7}}'{{else}}'{{invalid8}}'{{/if}};">
```
Comment (neraliu) [inline tests/samples/files/handlebarsjs_filter_attr_value_style_003.hbs]: please help to double check on this, as i found the single quote is being replaced by strict CP.

Comment (adon-at-work) [inline tests/samples/files/handlebarsjs_filter_attr_value_style_003.hbs]: Yes. single/double quotes in unquoted attr values are html escaped

File: src/css-parser.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+CSSParser.parseStyleAttributeValue = function(str) {
+    var r = { prop: '', code: NaN };
+    var kv = str.split(';'); // it will return new array even there is no ';' in the string
+    var v = kv[kv.length-1].split(':'); // only handling the last element
+    if (v.length === 2) {
+
+        // it is ok to trim the space at both ends
+        var prop = v[0].trim(), 
+            expr = v[1].trim();
+        if (prop === '') {
```
Comment (adon-at-work) [inline src/css-parser.js]: String.trim() is not supported by IE 8 or below
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/Trim

File: dist/secure-handlebars.js [hunk 7]
```diff
@@ -9744,7 +9832,202 @@ module.exports = ContextParserHandlebars;
...
+'&#0*32;?|&#[xX]0*20;?|&#0*9;?|&#[xX]0*9;?|&Tab;|&#0*10;?|&#[xX]0*[aA];?|&NewLine;|&#0*12;?|&#[xX]0*[cC];?|&#0*13;?|&#[xX]0*[dD];?|\t|\r|\n|\f'; // space,\t,\r,\n,\f
+
+note: the decoding order is important, as we are using regExp not real parsing. 
+for example, we need to replace \ before space, as &#0*92 needs to be checked before &#0*9.
+the correct decoding order is to match as greedy as possible first, that is matching N digit before N-1 digit.
+
+*/
+CSSParser.cssReplaceChar = [ '-', '\\', '_', ':', ';', '(', ')', '"', '\'', '\/', ',', '+', '%', '#', '!', '*', ' ' ];
+CSSParser.reCss = [
+    /&#0*8722;?|&#[xX]0*2212;?|&minus;|&#0*8208;?|&#[xX]0*2010;?|&dash;|&hyphen;/g,
```
Comment (adon-at-work) [inline dist/secure-handlebars.js]: `'-'.charCodeAt(0)` returns 45 (decimal). that hyphen symbol is copied from you `reExpr`. is 45 what you want to catch instead?

**Comment by neraliu** [comment]
the code review discussion has been done and this prototype is withdrawing with new approach.

**Pull request closed without merging**